common/containers/Tabs/SendTransaction/components/WalletInfo.tsx

```diff
@@ -0,0 +1,143 @@
+            </div>
+            <div className="col-xs-1" style={{ padding: 0 }}>
+              <Identicon address={address} />
+            </div>
+          </div>
+
+          {privateKey && (
+            <div className="row form-group">
+              <div className="col-xs-12">
+                <label>{translate('x_PrivKey')}</label>
+                <div className="input-group">
+                  <input
+                    className="form-control"
+                    disabled={true}
+                    type={isPrivateKeyVisible ? 'text' : 'password'}
+                    value={privateKey}
+                  />
+                  <span
+                    onClick={this.togglePrivateKey}
+                    aria-label={translateRaw('GEN_Aria_2')}
+                    role="button"
+                    className="input-group-addon eye"
+                  />
+                </div>
+              </div>
+            </div>
+          )}
+        </div>
+
+        <div className="Tab-content-pane">
+          <div className="row">
+            <div className="col-xs-6">
+              <label>Public Address</label>
+              <div className="WalletInfo-qr well well-lg">
+                <QRCode data={address} />
+              </div>
+            </div>
+            {privateKey && (
+              <div className="col-xs-6">
+                <label>Private Key</label>
+                <div className="WalletInfo-qr well well-lg" onClick={this.togglePrivateKey}>
+                  <QRCode data={isPrivateKeyVisible ? privateKey : '0'} />
+                  {!isPrivateKeyVisible && (
+                    <div className="WalletInfo-qr-cover">
+                      <i className="WalletInfo-qr-cover-icon fa fa-eye" />
+                    </div>
+                  )}
+                </div>
+              </div>
+            )}
+
+            {privateKey && (
+              <div className="col-xs-6">
+                <label>Utilities</label>
+
+                <button className="btn btn-info btn-block" onClick={print(address, privateKey)}>
+                  {translate('x_Print')}
+                </button>
+
+                <button className="btn btn-info btn-block" onClick={this.toggleKeystoreModal}>
+                  Generate Keystore File
+                </button>
+              </div>
+            )}
+
+            {isKeystoreModalOpen && (
+              <GenerateKeystoreModal
+                privateKey={privateKey}
+                handleClose={this.toggleKeystoreModal}
+              />
+            )}
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  private async setWalletAsyncState(wallet: IWallet) {
+    const address = await wallet.getAddressString();
+    let privateKey = '';
+
+    if (wallet.getPrivateKeyString) {
+      privateKey = await wallet.getPrivateKeyString();
+    }
+
+    this.setState({ address, privateKey });
+  }
+
+  private togglePrivateKey = () => {
```

Comment:
Same thoughts here regarding using the current state vs passing a callback to always get the correct previous state